File: packages/typescript/src/personality.ts
import {
  BigFiveConflictResolutionStyle,
  CONFLICT_RESOLUTION_STYLES,
  type BigFiveConflictResolutionStyle as ConflictStyle,
  type LifeStage,
  type PriorityLevel as Priority,
  PriorityLevel,
} from "./enums.js";
import {
  DEFAULT_RANDOM_SOURCE,
  type RandomOptions,
  type RandomSource,
} from "./randomness.js";
import {
  BigFiveAgreeableness,
  BigFiveConscientiousness,
  BigFiveExtraversion,
  BigFiveNeuroticism,
  BigFiveOpenness,
} from "./traits.js";

type TraitName =
  | "openness"
  | "conscientiousness"
  | "extraversion"
  | "agreeableness"
  | "neuroticism";
type TraitScores = Readonly<Record<TraitName, number>>;
type StyleCoefficients = Readonly<Record<TraitName, number>>;
const TRAIT_NAMES: readonly TraitName[] = Object.freeze([
  "openness",
  "conscientiousness",
  "extraversion",
  "agreeableness",
  "neuroticism",
]);

// Exported for compiled conformance tests; the package export map keeps this module internal.
export const CONFLICT_STYLE_COEFFICIENTS: Readonly<
  Record<ConflictStyle, StyleCoefficients>
> = Object.freeze({
  [BigFiveConflictResolutionStyle.Avoiding]: Object.freeze({
    openness: -0.1,
    conscientiousness: -0.2,
    extraversion: 0,
    agreeableness: 0.2,
    neuroticism: 0.7,
  }),
  [BigFiveConflictResolutionStyle.Obliging]: Object.freeze({
    openness: -0.1,
    conscientiousness: 0,
    extraversion: -0.2,
    agreeableness: 0.3,
    neuroticism: 0.2,
  }),
  [BigFiveConflictResolutionStyle.Integrating]: Object.freeze({
    openness: 0.1,
    conscientiousness: 0.1,
    extraversion: 0,
    agreeableness: 0.2,
    neuroticism: 0,
  }),
  [BigFiveConflictResolutionStyle.Dominating]: Object.freeze({
    openness: -0.2,
    conscientiousness: 0.2,
    extraversion: 0.2,
    agreeableness: -0.4,
    neuroticism: -0.2,
  }),
  [BigFiveConflictResolutionStyle.Compromising]: Object.freeze({
    openness: 0,
    conscientiousness: -0.2,
    extraversion: 0.1,
    agreeableness: 0,
    neuroticism: 0.1,
  }),
});

export const CONFLICT_MINIMUM_WEIGHT = 0.1;

export const CONFLICT_CONCERN_MAPPINGS: Readonly<
  Record<ConflictStyle, readonly [Priority, Priority]>
> = Object.freeze({
  [BigFiveConflictResolutionStyle.Avoiding]: Object.freeze([
    PriorityLevel.Low,
    PriorityLevel.Low,
  ] as const),
  [BigFiveConflictResolutionStyle.Obliging]: Object.freeze([
    PriorityLevel.Low,
    PriorityLevel.High,
  ] as const),
  [BigFiveConflictResolutionStyle.Integrating]: Object.freeze([
    PriorityLevel.High,
    PriorityLevel.High,
  ] as const),
  [BigFiveConflictResolutionStyle.Dominating]: Object.freeze([
    PriorityLevel.High,
    PriorityLevel.Low,
  ] as const),
  [BigFiveConflictResolutionStyle.Compromising]: Object.freeze([
    PriorityLevel.Moderate,
    PriorityLevel.Moderate,
  ] as const),
});

function traitScores(traits: BigFiveTraits): TraitScores {
  return {
    openness: traits.openness.score,
    conscientiousness: traits.conscientiousness.score,
    extraversion: traits.extraversion.score,
    agreeableness: traits.agreeableness.score,
    neuroticism: traits.neuroticism.score,
  };
}

export function conflictStyleWeights(
  traits: BigFiveTraits,
): Readonly<Record<ConflictStyle, number>> {
  const scores = traitScores(traits);
  const weights = {} as Record<ConflictStyle, number>;
  for (const style of CONFLICT_RESOLUTION_STYLES) {
    const coefficients = CONFLICT_STYLE_COEFFICIENTS[style];
    const level = TRAIT_NAMES.reduce(
      (total, traitName) => total + scores[traitName] * coefficients[traitName],
      0,
    );
    weights[style] = Math.max(level, CONFLICT_MINIMUM_WEIGHT);
  }
  return Object.freeze(weights);
}

function weightedChoice(
  weights: Readonly<Record<ConflictStyle, number>>,
  rng: RandomSource,
): ConflictStyle {
  const total = CONFLICT_RESOLUTION_STYLES.reduce(
    (sum, style) => sum + weights[style],
    0,
  );
  const draw = rng.uniform(0, total);
  if (!Number.isFinite(draw) || draw < 0 || draw > total) {
    throw new RangeError("RandomSource.uniform returned a value out of range.");
  }
  let threshold = draw;
  for (const style of CONFLICT_RESOLUTION_STYLES) {
    threshold -= weights[style];
    if (threshold <= 0) {
      return style;
    }
  }
  return CONFLICT_RESOLUTION_STYLES[
    CONFLICT_RESOLUTION_STYLES.length - 1
  ] as ConflictStyle;
}

export interface BigFiveTraitsOptions {
  readonly openness: BigFiveOpenness;
  readonly conscientiousness: BigFiveConscientiousness;
  readonly extraversion: BigFiveExtraversion;
  readonly agreeableness: BigFiveAgreeableness;
  readonly neuroticism: BigFiveNeuroticism;
}

export class BigFiveTraits {
  readonly openness: BigFiveOpenness;
  readonly conscientiousness: BigFiveConscientiousness;
  readonly extraversion: BigFiveExtraversion;
  readonly agreeableness: BigFiveAgreeableness;
  readonly neuroticism: BigFiveNeuroticism;

  constructor(options: BigFiveTraitsOptions) {
    if (
      !(options.openness instanceof BigFiveOpenness) ||
      !(options.conscientiousness instanceof BigFiveConscientiousness) ||
      !(options.extraversion instanceof BigFiveExtraversion) ||
      !(options.agreeableness instanceof BigFiveAgreeableness) ||
      !(options.neuroticism instanceof BigFiveNeuroticism)
    ) {
      throw new TypeError("Big Five traits must be Big Five trait objects.");
    }
    this.openness = options.openness;
    this.conscientiousness = options.conscientiousness;
    this.extraversion = options.extraversion;
    this.agreeableness = options.agreeableness;
    this.neuroticism = options.neuroticism;
    Object.freeze(this);
  }

  static random(lifeStage: LifeStage, options: RandomOptions = {}): BigFiveTraits {
    return new BigFiveTraits({
      openness: BigFiveOpenness.random(lifeStage, options),
      conscientiousness: BigFiveConscientiousness.random(lifeStage, options),
      extraversion: BigFiveExtraversion.random(lifeStage, options),
      agreeableness: BigFiveAgreeableness.random(lifeStage, options),
      neuroticism: BigFiveNeuroticism.random(lifeStage, options),
    });
  }
}

export interface BigFiveConflictResolutionOptions {
  readonly style: ConflictStyle;
}

export class BigFiveConflictResolution {
  readonly style: ConflictStyle;
  readonly concernForSelf: Priority;
  readonly concernForOthers: Priority;

  constructor(options: BigFiveConflictResolutionOptions) {
    if (!CONFLICT_RESOLUTION_STYLES.includes(options.style)) {
      throw new TypeError(`Unsupported conflict style: ${String(options.style)}`);
    }
    const [concernForSelf, concernForOthers] = CONFLICT_CONCERN_MAPPINGS[options.style];
    this.style = options.style;
    this.concernForSelf = concernForSelf;
    this.concernForOthers = concernForOthers;
    Object.freeze(this);
  }

  static random(
    traits: BigFiveTraits,
    options: RandomOptions = {},
  ): BigFiveConflictResolution {
    const style = weightedChoice(
      conflictStyleWeights(traits),
      options.rng ?? DEFAULT_RANDOM_SOURCE,
    );
    return new BigFiveConflictResolution({
      style,
    });
  }
}

export interface BigFivePersonalityOptions {
  readonly traits: BigFiveTraits;
  readonly conflictResolution: BigFiveConflictResolution;
}

export class BigFivePersonality {
  readonly traits: BigFiveTraits;
  readonly conflictResolution: BigFiveConflictResolution;

  constructor(options: BigFivePersonalityOptions) {
    if (!(options.traits instanceof BigFiveTraits)) {
      throw new TypeError("traits must be BigFiveTraits.");
    }
    if (!(options.conflictResolution instanceof BigFiveConflictResolution)) {
      throw new TypeError("conflictResolution must be a BigFiveConflictResolution.");
    }
    this.traits = options.traits;
    this.conflictResolution = options.conflictResolution;
    Object.freeze(this);
  }

  static random(lifeStage: LifeStage, options: RandomOptions = {}): BigFivePersonality {
    const traits = BigFiveTraits.random(lifeStage, options);
    return new BigFivePersonality({
      traits,
      conflictResolution: BigFiveConflictResolution.random(traits, options),
    });
  }
}

File: packages/python/src/personalitygen/personality.py
"""Top-level personality configuration models."""

from __future__ import annotations

import math
import random
from dataclasses import dataclass, field
from enum import Enum
from typing import Self

from personalitygen.enums import LifeStage, PriorityLevel
from personalitygen.randomness import RandomSource
from personalitygen.traits import (
    BigFiveAgreeableness,
    BigFiveConscientiousness,
    BigFiveExtraversion,
    BigFiveNeuroticism,
    BigFiveOpenness,
)


def _weighted_choice(
    weights: dict["BigFiveConflictResolutionStyle", float],
    *,
    rng: RandomSource | None = None,
) -> "BigFiveConflictResolutionStyle":
    if not weights:
        raise ValueError("weights must be non-empty")
    if any(
        not math.isfinite(weight) or weight < 0.0
        for weight in weights.values()
    ):
        raise ValueError("weights must be finite and non-negative")
    source = rng if rng is not None else random
    total = sum(weights.values())
    if total <= 0.0:
        weights = {style: 1.0 for style in weights}
        total = float(len(weights))

    threshold = source.uniform(0.0, total)
    if (
        not math.isfinite(threshold)
        or threshold < 0.0
        or threshold > total
    ):
        raise ValueError("RandomSource.uniform returned a value out of range")
    for style, weight in weights.items():
        threshold -= weight
        if threshold <= 0.0:
            return style
    return next(reversed(weights))


class BigFiveConflictResolutionStyle(str, Enum):
    # Concern for self: low. Concern for others: low. Tries to avoid conflict.
    AVOIDING = "avoiding"
    # Concern for self: low. Concern for others: high. Accommodates others.
    OBLIGING = "obliging"
    # Concern for self: high. Concern for others: high. Collaborates.
    INTEGRATING = "integrating"
    # Concern for self: high. Concern for others: low. Competes to win.
    DOMINATING = "dominating"
    # Concern for self: moderate. Concern for others: moderate. Trades off.
    COMPROMISING = "compromising"

    @classmethod
    def random(
        cls,
        traits: BigFiveTraits,
        *,
        rng: RandomSource | None = None,
    ) -> Self:
        return _weighted_choice(
            _conflict_style_weights(traits),
            rng=rng,
        )


@dataclass(frozen=True, slots=True)
class BigFiveTraits:
    # Appreciation for art, emotion, adventure, and curiosity.
    # Opposite: closedness.
    openness: BigFiveOpenness
    # Self-discipline, dutifulness, and achievement orientation.
    # Opposite: undisciplined.
    conscientiousness: BigFiveConscientiousness
    # Energy, sociability, and stimulation-seeking.
    # Opposite: introversion.
    extraversion: BigFiveExtraversion
    # Compassion and cooperation toward others.
    # Opposite: antagonism.
    agreeableness: BigFiveAgreeableness
    # Tendency toward unpleasant emotions and instability.
    # Opposite: emotional stability.
    neuroticism: BigFiveNeuroticism

    @classmethod
    def random(
        cls, life_stage: LifeStage, *, rng: RandomSource | None = None
    ) -> Self:
        return cls(
            openness=BigFiveOpenness.random(life_stage, rng=rng),
            conscientiousness=BigFiveConscientiousness.random(
                life_stage, rng=rng
            ),
            extraversion=BigFiveExtraversion.random(life_stage, rng=rng),
            agreeableness=BigFiveAgreeableness.random(life_stage, rng=rng),
            neuroticism=BigFiveNeuroticism.random(life_stage, rng=rng),
        )

    def __str__(self) -> str:
        return (
            "openness: "
            f"{self.openness}\n"
            "conscientiousness: "
            f"{self.conscientiousness}\n"
            "extraversion: "
            f"{self.extraversion}\n"
            "agreeableness: "
            f"{self.agreeableness}\n"
            "neuroticism: "
            f"{self.neuroticism}"
        )


# These weights are loosely based on:
# Priyadarshini, S. (2017). Effect of Personality on Conflict Resolution
# Styles. IRA-International Journal of Management & Social Sciences, 7(2),
# 196-207.
_CONFLICT_STYLE_COEFFICIENTS: dict[
    BigFiveConflictResolutionStyle, dict[str, float]
] = {
    BigFiveConflictResolutionStyle.AVOIDING: {
        "neuroticism": 0.7,
        "openness": -0.1,
        "agreeableness": 0.2,
        "conscientiousness": -0.2,
        "extraversion": 0.0,
    },
    BigFiveConflictResolutionStyle.OBLIGING: {
        "neuroticism": 0.2,
        "extraversion": -0.2,
        "openness": -0.1,
        "agreeableness": 0.3,
        "conscientiousness": 0.0,
    },
    BigFiveConflictResolutionStyle.INTEGRATING: {
        "openness": 0.1,
        "agreeableness": 0.2,
        "conscientiousness": 0.1,
        "neuroticism": 0.0,
        "extraversion": 0.0,
    },
    BigFiveConflictResolutionStyle.DOMINATING: {
        "neuroticism": -0.2,
        "extraversion": 0.2,
        "openness": -0.2,
        "agreeableness": -0.4,
        "conscientiousness": 0.2,
    },
    BigFiveConflictResolutionStyle.COMPROMISING: {
        "neuroticism": 0.1,
        "extraversion": 0.1,
        "conscientiousness": -0.2,
        "openness": 0.0,
        "agreeableness": 0.0,
    },
}
_CONFLICT_MINIMUM_WEIGHT = 0.1


def _conflict_style_weights(
    traits: BigFiveTraits,
) -> dict[BigFiveConflictResolutionStyle, float]:
    trait_scores = {
        "openness": traits.openness.score,
        "conscientiousness": traits.conscientiousness.score,
        "extraversion": traits.extraversion.score,
        "agreeableness": traits.agreeableness.score,
        "neuroticism": traits.neuroticism.score,
    }
    return {
        style: max(
            sum(
                trait_scores[trait_name] * coefficient
                for trait_name, coefficient in coefficients.items()
            ),
            _CONFLICT_MINIMUM_WEIGHT,
        )
        for style, coefficients in _CONFLICT_STYLE_COEFFICIENTS.items()
    }


_STYLE_TO_CONCERNS: dict[
    BigFiveConflictResolutionStyle, tuple[PriorityLevel, PriorityLevel]
] = {
    BigFiveConflictResolutionStyle.AVOIDING: (
        PriorityLevel.LOW,
        PriorityLevel.LOW,
    ),
    BigFiveConflictResolutionStyle.OBLIGING: (
        PriorityLevel.LOW,
        PriorityLevel.HIGH,
    ),
    BigFiveConflictResolutionStyle.INTEGRATING: (
        PriorityLevel.HIGH,
        PriorityLevel.HIGH,
    ),
    BigFiveConflictResolutionStyle.DOMINATING: (
        PriorityLevel.HIGH,
        PriorityLevel.LOW,
    ),
    BigFiveConflictResolutionStyle.COMPROMISING: (
        PriorityLevel.MODERATE,
        PriorityLevel.MODERATE,
    ),
}


def _validate_style_concerns() -> None:
    expected = set(BigFiveConflictResolutionStyle)
    actual = set(_STYLE_TO_CONCERNS)
    if expected != actual:
        missing = {style.value for style in expected - actual}
        extra = {style.value for style in actual - expected}
        raise ValueError(
            "Conflict resolution styles and concern mapping are out of sync. "
            f"Missing: {sorted(missing)}. Extra: {sorted(extra)}."
        )


_validate_style_concerns()


@dataclass(frozen=True, slots=True)
class BigFiveConflictResolution:
    style: BigFiveConflictResolutionStyle
    concern_for_self: PriorityLevel = field(init=False)
    concern_for_others: PriorityLevel = field(init=False)

    def __post_init__(self) -> None:
        if not isinstance(
            self.style,
            BigFiveConflictResolutionStyle,
        ):
            raise TypeError(
                "style must be a BigFiveConflictResolutionStyle"
            )
        concern_for_self, concern_for_others = _STYLE_TO_CONCERNS[
            self.style
        ]
        object.__setattr__(self, "concern_for_self", concern_for_self)
        object.__setattr__(self, "concern_for_others", concern_for_others)

    @classmethod
    def random(
        cls,
        traits: BigFiveTraits,
        *,
        rng: RandomSource | None = None,
    ) -> Self:
        style = BigFiveConflictResolutionStyle.random(
            traits, rng=rng
        )
        return cls(style=style)


@dataclass(frozen=True, slots=True)
class BigFivePersonality:
    traits: BigFiveTraits
    conflict_resolution: BigFiveConflictResolution

    @classmethod
    def random(
        cls, life_stage: LifeStage, *, rng: RandomSource | None = None
    ) -> Self:
        traits = BigFiveTraits.random(life_stage, rng=rng)
        conflict_resolution = BigFiveConflictResolution.random(
            traits, rng=rng
        )
        return cls(
            traits=traits,
            conflict_resolution=conflict_resolution,
        )
